ublox_derive/src/output.rs

```diff
@@ -0,0 +1,572 @@
+    let mut items = Vec::with_capacity(usize::try_from(bitflags.nbits).unwrap());
+    let repr_ty = &bitflags.repr_ty;
+
+    for bit in 0..bitflags.nbits {
+        let flag_bit = 1u64.checked_shl(bit).unwrap();
+        if let Some(item) = bitflags.consts.iter().find(|x| x.value == flag_bit) {
+            known_flags.insert(flag_bit);
+            let name = &item.name;
+            let attrs = &item.attrs;
+            if bit != 0 {
+                items.push(quote! {
+                    #(#attrs)*
+                    const #name  = ((1 as #repr_ty) << #bit)
+                });
+            } else {
+                items.push(quote! {
+                    #(#attrs)*
+                    const #name  = (1 as #repr_ty)
+                });
+            }
+        } else {
+            let name = format_ident!("RESERVED{}", bit);
+            if bit != 0 {
+                items.push(quote! { const #name = ((1 as #repr_ty) << #bit)});
+            } else {
+                items.push(quote! { const #name = (1 as #repr_ty) });
+            }
+        }
+    }
+
+    if known_flags.len() != bitflags.consts.len() {
+        let user_flags: HashSet<_> = bitflags.consts.iter().map(|x| x.value).collect();
+        let set = user_flags.difference(&known_flags);
+        return Err(syn::Error::new(
+            bitflags.name.span(),
+            format!("Strange flags, not power of 2?: {:?}", set),
+        ));
+    }
+
+    let vis = &bitflags.vis;
+    let attrs = &bitflags.attrs;
+    let name = &bitflags.name;
+
+    let from = match bitflags.from_fn {
+        None => quote! {},
+        Some(UbxTypeFromFn::From) => quote! {
+            impl #name {
+                const fn from(x: #repr_ty) -> Self {
+                    Self::from_bits_truncate(x)
+                }
+            }
+        },
+        Some(UbxTypeFromFn::FromUnchecked) => unimplemented!(),
+    };
+
+    let into = match bitflags.into_fn {
+        None => quote! {},
+        Some(UbxTypeIntoFn::Raw) => quote! {
+            impl #name {
+                const fn into_raw(self) -> #repr_ty {
+                    self.bits()
+                }
+            }
+        },
+    };
+
+    Ok(quote! {
+        bitflags! {
+            #(#attrs)*
+            #vis struct #name : #repr_ty {
+                #(#items);*;
+            }
+        }
+        #from
+        #into
+    })
+}
+
+pub fn generate_code_for_parse(recv_packs: &RecvPackets) -> TokenStream {
+    let mut pack_enum_variants = vec![];
+    let mut matches = vec![];
+    let mut max_16_chain = None;
+
+    for name in &recv_packs.all_packets {
+        let ref_name = format_ident!("{}Ref", name);
+        pack_enum_variants.push(quote! {
+            #name(#ref_name <'a>)
+        });
+
+        matches.push(quote! {
+            (#name::CLASS, #name::ID) if <#ref_name>::validate(payload).is_ok()  => {
+                Ok(PacketRef::#name(#ref_name(payload)))
+            }
+        });
```

Comment:
I think these could each be generated from maps.

Reply:
It would be more complicated then existing code,
something like this:

```
    let (x, y): (Vec<_>, Vec<_>) = recv_packs
        .all_packets
        .iter()
        .map(|name| {
            let ref_name = format_ident!("{}Ref", name);
            (
                quote! {
                    #name(#ref_name <'a>)
                },
                quote! {
                    (#name::CLASS, #name::ID) if <#ref_name>::validate(payload).is_ok()  => {
                        Ok(PacketRef::#name(#ref_name(payload)))
                    }
                },
            )
        })
        .unzip();
```